flink/v1.15/flink/src/main/java/org/apache/iceberg/flink/sink/v2/IcebergFilesCommitter.java

```diff
@@ -0,0 +1,261 @@
+  private final boolean replacePartitions;
+  private final Map<String, String> snapshotProperties;
+
+  // It will have an unique identifier for one job.
+  private final Table table;
+  private transient long maxCommittedCheckpointId;
+  private transient int continuousEmptyCheckpoints;
+  private transient int maxContinuousEmptyCommits;
+
+  private transient ExecutorService workerPool;
+
+  public IcebergFilesCommitter(TableLoader tableLoader,
+                               boolean replacePartitions,
+                               Map<String, String> snapshotProperties,
+                               Integer workerPoolSize) {
+
+    this.tableLoader = tableLoader;
+    this.replacePartitions = replacePartitions;
+    this.snapshotProperties = snapshotProperties;
+
+    this.workerPool =
+        ThreadPools.newWorkerPool("iceberg-worker-pool-" + Thread.currentThread().getName(), workerPoolSize);
+
+    // Open the table loader and load the table.
+    this.tableLoader.open();
+    this.table = tableLoader.loadTable();
+
+    maxContinuousEmptyCommits = PropertyUtil.propertyAsInt(table.properties(),
+        MAX_CONTINUOUS_EMPTY_COMMITS,
+        MAX_CONTINUOUS_EMPTY_COMMITS_DEFAULT);
+    Preconditions.checkArgument(
+        maxContinuousEmptyCommits > 0,
+        MAX_CONTINUOUS_EMPTY_COMMITS + " must be positive");
+    this.maxCommittedCheckpointId = INITIAL_CHECKPOINT_ID;
+  }
+
+  @Override
+  public void commit(Collection<CommitRequest<IcebergFlinkCommittable>> requests)
+      throws IOException, InterruptedException {
+    int dataFilesNumRestored = 0;
+    int deleteFilesNumRestored = 0;
+
+    int dataFilesNumStored = 0;
+    int deleteFilesNumStored = 0;
+
+    long checkpointId = 0;
+    Collection<CommitRequest<IcebergFlinkCommittable>> restored = Lists.newArrayList();
+    Collection<CommitRequest<IcebergFlinkCommittable>> store = Lists.newArrayList();
+
+    for (CommitRequest<IcebergFlinkCommittable> request : requests) {
+      checkpointId = request.getCommittable().checkpointId();
+      WriteResult committable = request.getCommittable().committable();
+      if (request.getNumberOfRetries() > 0) {
+        if (request.getCommittable().checkpointId() > maxCommittedCheckpointId) {
+          restored.add(request);
+          dataFilesNumRestored = dataFilesNumRestored + committable.dataFiles().length;
+          deleteFilesNumRestored = deleteFilesNumRestored + committable.deleteFiles().length;
+        }
+      } else {
+        if (request.getCommittable().checkpointId() > maxCommittedCheckpointId) {
+          store.add(request);
+          dataFilesNumStored = dataFilesNumStored + committable.dataFiles().length;
+          deleteFilesNumStored = deleteFilesNumStored + committable.deleteFiles().length;
+        }
+      }
+    }
+
+    if (restored.size() > 0) {
+      commitResult(dataFilesNumRestored, deleteFilesNumRestored, restored);
+    }
+    commitResult(dataFilesNumStored, deleteFilesNumStored, store);
```

Comment:
Do we always have new stuff to store? Are the retries handled separately?

Reply:
Not necessarily. When we have no data to commit, we increase the number of empty commits by one. When it is greater than the threshold, we also make a commit.
```java
   private void commitResult(int dataFilesNum, int deleteFilesNum,
      Collection<CommitRequest<FilesCommittable>> committableCollection) {
    int totalFiles = dataFilesNum + deleteFilesNum;

    continuousEmptyCheckpoints = totalFiles == 0 ? continuousEmptyCheckpoints + 1 : 0;
    if (totalFiles != 0 || continuousEmptyCheckpoints % maxContinuousEmptyCommits == 0) {
      if (replacePartitions) {
        replacePartitions(committableCollection, deleteFilesNum);
      } else {
        commitDeltaTxn(committableCollection, dataFilesNum, deleteFilesNum);
      }
      continuousEmptyCheckpoints = 0;
    }
  }
```

Reply:
If I understand correctly we count the empty checkpoints, but why do we need to create a new commit after 10 checkpoints, even if we did not write anything? Is this something like a liveliness check? Does this update some metadata on the table, even if no new files are added?

Thanks!
Peter

Reply:
We will have a new snapshot with the same manifests as the old snapshot.
Nice catch! 
The reason for this logic is to refresh the latest `max-committed-checkpoint-id`, so that it can correctly recover from state. Now we change state to `FilesCommittable` and leave it to the Flink itself. So it looks like we don't need that logic anymore.